fix(perps): enforce 5 significant figure limit for TP/SL price inputs

michalconsensys opened the pull request:
This PR addresses HyperLiquid API rejections that occur when TP/SL prices exceed 5 significant figures.

**Problem:**
- HyperLiquid API rejects orders with prices containing more than 5 significant figures
- Users could type prices with excessive precision that would fail on submission
- RoE percentage calculations could produce prices exceeding the limit

**Solution:**
1. Added input validation to prevent users from typing more than 5 significant figures in TP/SL price fields
2. Added rounding logic when prices are calculated from RoE percentages to ensure they stay within limits
3. Added `MAX_SIGNIFICANT_FIGURES` constant for centralized configuration
4. Fixed TP/SL price display in position cards to use consistent formatting (`PRICE_RANGES_UNIVERSAL`)

**Technical changes:**
- Added three new utility functions: `countSignificantFigures`, `hasExceededSignificantFigures`, `roundToSignificantFigures`
- Fixed regex bug in `countSignificantFigures` that incorrectly handled negative numbers (`/^-?0+/` → `/^-?0*/`)
- Added comprehensive unit tests for the new validation utilities


## **Changelog**

CHANGELOG entry: Fixed TP/SL price input validation to respect HyperLiquid's 5 significant figure limit


## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/TAT-2063

## **Manual testing steps**

```gherkin
Feature: TP/SL significant figure validation

  Scenario: user cannot type more than 5 significant figures
    Given user has an open perps position
    And user opens the TP/SL form

    When user types "123.45" in the Take Profit price field
    Then the input accepts the value (5 sig figs)
    
    When user tries to type "123.456" in the Take Profit price field
    Then the input blocks the 6th significant figure

  Scenario: RoE percentage calculation rounds to 5 significant figures
    Given user has an open perps position
    And user opens the TP/SL form

    When user enters a RoE percentage that would calculate to "123.456789"
    Then the price field displays the value rounded to 5 significant figures
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

User was able to enter more decimals than supported

### **After**


https://github.com/user-attachments/assets/9c563850-dcfd-4880-934d-5e0efd6953be



## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Ensures TP/SL prices comply with HyperLiquid’s 5 significant-figure rule and aligns formatting across UI and adapters.
> 
> - Introduces `DECIMAL_PRECISION_CONFIG.MAX_SIGNIFICANT_FIGURES` and shared utils: `countSignificantFigures`, `hasExceededSignificantFigures`, `roundToSignificantFigures`
> - Updates `usePerpsTPSLForm` to prevent typing beyond 5 sig figs and round prices derived from RoE inputs/buttons; maintains bidirectional sync without interfering while fields are focused
> - Switches TP/SL display in `PerpsPositionCard` to `PRICE_RANGES_UNIVERSAL` for consistent price formatting
> - Refactors `formatHyperLiquidPrice` in `hyperLiquidAdapter` to use shared significant-figure rounding logic
> - Adds comprehensive unit tests for formatting/validation and adjusts existing tests to match new rounding behavior
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 59a6a650520aeb6a5fb9796d263612047daec1b2. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

michalconsensys committed:
feat(perps): add precision warning for TP/SL price inputs exceeding 5 significant figures

michalconsensys committed:
fix(perps): improve TP/SL precision warning to match HyperLiquid rounding behavior

michalconsensys committed:
change range to universal

michalconsensys committed:
test(perps): add unit tests for significant figures validation utilities

michalconsensys committed:
Merge branch 'main' into fix/perps/digits-warning

michalconsensys committed:
chore: do not allow to enter numbers exceeding significant figures

michalconsensys committed:
chore: add rounding

michalconsensys committed:
chore: improve roundToSignificantFigures

michalconsensys committed:
Merge branch 'fix/perps/digits-warning' of github.com:MetaMask/metamask-mobile into fix/perps/digits-warning

michalconsensys committed:
chore: fix negative numbers

michalconsensys committed:
chore: remove unused translation

michalconsensys commented on app/components/UI/Perps/utils/tpslValidation.ts:
```diff
@@ -13,6 +13,84 @@
+  const [integerPart, decimalPart = ''] = normalized.split('.');
+
+  // Remove leading zeros and negative sign from integer part
+  const trimmedInteger = integerPart.replace(/^-?0+/, '') || '';
```
Fixed in https://github.com/MetaMask/metamask-mobile/pull/24169/commits/157a784774db45c969857008e1f1d9bba2632ecb

gambinish reviewed (changes requested):
It looks like the PR includes new utility functions to prevent a user from entering more than 5 significant digits when inputing TP/SL. But, we're still supporting asset pairs with more than 5 significant digits like `MON` `PENGU` `PUMP` etc.

So I run into a situation where I can't set a TP/SL to the same precision as the listed asset pair.

Are we sure that we only want to support 5 significant digits? According to the docs, it appears that we can actually support more than 5 significant digits in certain cases. https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/tick-and-lot-size#perp-price-examples

Adding this comment here for viz, but I'm happy to chat about this more in Slack

gambinish committed:
Merge branch 'main' into fix/perps/digits-warning

gambinish commented:
Actually, after reviewing this again, I think that the input validation is working as it should. It is actually valid to not be able to add a TP/SL to the same level of precision as the actual asset pair (still seems weird to me)

I'm ready to approve this, but do think that the bugbot comments are legitimate

michalconsensys committed:
fix: update tests

michalconsensys committed:
test(perps): add unit tests for roundToSignificantFigures function

michalconsensys committed:
fix(perps): align regex in roundToSignificantFigures with countSignificantFigures

Change regex from /^-?0+/ to /^-?0*/ to properly strip minus sign
when counting significant figures for negative numbers. The 0+
quantifier required at least one zero to match, causing the minus
sign to remain in trimmedInteger for numbers like -12.345.

michalconsensys committed:
Merge branch 'fix/perps/digits-warning' of github.com:MetaMask/metamask-mobile into fix/perps/digits-warning

michalconsensys committed:
fix(perps): round prices in percentage text input handlers to 5 sig figs

michalconsensys committed:
fix(perps): apply significant figures rounding in blur handlers

Add roundToSignificantFigures to handleTakeProfitPercentageBlur,
handleStopLossPercentageBlur, and handleStopLossPriceBlur (zero ROE case)
to ensure consistency with other handlers and prevent HyperLiquid API
rejection due to exceeding 5 significant figures.

michalconsensys commented on app/components/UI/Perps/hooks/usePerpsTPSLForm.ts:
```diff
@@ -497,7 +513,9 @@ export function usePerpsTPSLForm(
-        setStopLossPrice(price.toString());
+        // Round to 5 significant figures to match input validation
+        const roundedPrice = roundToSignificantFigures(price.toString());
+        setStopLossPrice(roundedPrice);
```
Resolved in https://github.com/MetaMask/metamask-mobile/pull/24169/commits/b5af764e09b6f344250a8e08e73e126b09a369a8

michalconsensys committed:
fix(perps): use >= comparison for significant figures validation

Change from > to >= in TPSL form handlers to prevent bypassing
the 5 significant figure limit when pasting same-length values.
This makes the check consistent with the decimal places validation.

michalconsensys committed:
refactor(perps): consolidate significant figures utilities into formatUtils

- Move countSignificantFigures, hasExceededSignificantFigures, and
  roundToSignificantFigures from tpslValidation.ts to formatUtils.ts
- Update formatHyperLiquidPrice to use shared utilities instead of
  inline logic
- Re-export utilities from tpslValidation.ts for backwards compatibility
- Move related tests from tpslValidation.test.ts to formatUtils.test.ts
- Fix test mock in usePerpsTPSLForm.test.ts to include new exports

michalconsensys committed:
fix(perps): add missing significant figures rounding in handleTakeProfitPriceBlur

michalconsensys committed:
Merge branch 'main' into fix/perps/digits-warning

michalconsensys committed:
Merge branch 'main' into fix/perps/digits-warning

michalconsensys committed:
Merge branch 'main' into fix/perps/digits-warning
